src/python/pants/rules/core/strip_source_roots_test.py

```diff
@@ -1,55 +1,97 @@
 # Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
-from typing import Optional
+from functools import partial
+from typing import List, Optional, Union
 from unittest.mock import Mock
 
-from pants.build_graph.address import Address
+import pytest
+
 from pants.build_graph.files import Files
 from pants.engine.legacy.graph import HydratedTarget
 from pants.engine.rules import RootRule
+from pants.engine.scheduler import ExecutionError
 from pants.engine.selectors import Params
-from pants.rules.core.strip_source_root import SourceRootStrippedSources
-from pants.rules.core.strip_source_root import rules as strip_source_root_rules
+from pants.rules.core.strip_source_roots import SnapshotToStrip, SourceRootStrippedSources
+from pants.rules.core.strip_source_roots import rules as strip_source_root_rules
 from pants.testutil.option.util import create_options_bootstrapper
 from pants.testutil.test_base import TestBase
 
 
 class StripSourceRootsTests(TestBase):
   @classmethod
   def rules(cls):
-    return (*super().rules(), *strip_source_root_rules(), RootRule(HydratedTarget))
-
-  def assert_stripped_source_file(
-    self, *, original_path: str, expected_path: str, target_type_alias: Optional[str] = None,
-  ) -> None:
-    adaptor = Mock()
-    adaptor.sources = Mock()
-    adaptor.sources.snapshot = self.make_snapshot({original_path: "print('random python')"})
-    adaptor.address = Mock()
-    adaptor.address.spec_path = original_path
-    if target_type_alias:
-      adaptor.type_alias = target_type_alias
-    target = HydratedTarget(Address.parse("some/target:target"), adaptor, ())
-    stripped_sources = self.request_single_product(
-      SourceRootStrippedSources, Params(target, create_options_bootstrapper())
+    return (
+      *super().rules(),
+      *strip_source_root_rules(),
+      RootRule(HydratedTarget),
+      RootRule(SnapshotToStrip),
     )
-    self.assertEqual(stripped_sources.snapshot.files, (expected_path,))
 
-  def test_source_roots_python(self):
-    self.assert_stripped_source_file(
-      original_path='src/python/pants/util/strutil.py', expected_path='pants/util/strutil.py',
+  def get_stripped_files(
+    self, original: Union[SnapshotToStrip, HydratedTarget], *, args: Optional[List[str]] = None,
+  ) -> List[str]:
+    result = self.request_single_product(
+      SourceRootStrippedSources, Params(original, create_options_bootstrapper(args=args))
     )
+    return sorted(result.snapshot.files)
 
-  def test_source_roots_java(self):
-    self.assert_stripped_source_file(
-      original_path='src/java/some/path/to/something.java',
-      expected_path='some/path/to/something.java',
-    )
+  def test_strip_snapshot(self) -> None:
+    original_paths = [
+      "src/python/project/example.py",
+      "src/java/com/project/example.java",
+      "tests/python/project_test/example.py",
+      "no-source-root/example.txt",
+    ]
```

Comment:
The most important question of this PR is whether we care about supporting the edge cases of a `Snapshot` or a target's `sources` field having files with different source roots?

When might this happen? For targets, when you have a [target aggregate](https://www.pantsbuild.org/target_aggregate.html). For filesystem specs, when you use a very loose glob like `./src python test '**/*.py'`.

--

Conceptually, it is the "correct" thing to do to support this use case. The point of these rules are to strip all source roots, regardless of whether the source roots look like their neighbors.

Concretely, without this feature, if a user ever had this case where they tried mixing files with different source roots, then the engine would fail because it would try, for example, stripping `src/python` from the file `tests/python/pants_test/example.py`.

--

Why not do this? It makes the implementation more complex and almost certainly less performant. 

If we didn't care about this edge case, then we would change the `SnapshotToStrip` class to have an additional field called something like `example_file`. We would determine one single prefix to strip for the entire collection, exclusively determined by the `example_file`. For targets, `example_file` would be the target address. For snapshots, it'd be a random file, likely.

Compared to this simpler implementation (which is what we had before this PR), this PR's implementation has a few performance implications:

* We call `SourceRoots.find_by_path` on _every_ source file used transitively for goals like `test`, compared to calling it once per every target used transitively.
* We must call `groupby` on every single source file used transitively.
* We must call `SnapshotSubset` and `DirectoriesToMerge`.

I haven't profiled, but things _felt_ slower with this PR and the source root stripping showed up in `--v2-ui`.

--

If we decide we do care about supporting this edge case, then we can possibly make these optimizations:

1. If there is only one source root for the `Snapshot`, then early return without needing to use `SnapshotSubset` and `DirectoriesToMerge`.
1. Memoize `SourcesRoot.find_by_path`?

Reply:
> When might this happen? For targets, when you have a target aggregate. For filesystem specs, when you use a very loose glob like ./src python test '**/*.py'.

Hm, I realized this second case with args is not actually true, as explained below. Please disregard it.

When given a file glob like `**/*.py`, each resulting `AddressWithOrigin` will have in its `origin` field an instance of  the same `FilesystemResolvedGlobSpec`, where the `resolved_files` refer to every single Python file in `**/*.py`, i.e. more than the `sources` belonging to that target.

For example, let's say we run `./pants test 'tests/python/pants_test/util/test_*.py'`, which resolves to multiple targets like `:strutil` and `:dirutil`. Each `AddressWithOrigin` will still refer to every single `util/test_*.py` file in the `origin`, even if some of those files aren't owned by the corresponding `Address`. It would be wrong to pass to the Pytest CLI every single resolved file, because that would mean trying to run every single `util/test_*.py` file for one single target like `:strutil`. Instead, when testing `:strutil`, we should only pass to Pytest the file `test_strutil.py`, even though the resolved glob was more expansive.

In other words, when dealing with filesystem specs, the file names we pass to Pytest must always be a subset of the original `sources` for the owning target.

--

So, the question really is: **how much do we care about supporting targets whose `sources` field has multiple source roots in it?**

That seems very unlikely to happen, outside of target aggregates. Generally, target aggregates have the type `target` rather than something like `python_library` or `python_tests`, so in the case of `fmt`, `lint`, and `test`, we would ignore that target aggregate to begin with.

After writing this all I out, I do think we should go back to only looking at one single file to determine the prefix for the entire collection of files, almost entirely on the grounds of performance. These rules get called extremely frequently (once for every single target used transitively), so a slight performance regression for each invocation does add up. 

Let me know if I'm missing something / you disagree.

Reply:
I don't think targets can have sources from multiple source roots. Or rather, while nothing specifically enforces this, I believe, v1 targets have a target_base property (https://github.com/pantsbuild/pants/blob/388d9e839736caf160ef8fe0d14e3c505fdda392/src/python/pants/build_graph/target.py#L171) that makes no sense unless the target is in a single source root. 

So I think we can actually error if we encounter such a target.